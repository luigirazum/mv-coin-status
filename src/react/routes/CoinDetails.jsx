import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import genericCoinIcon from '../../assets/logo/generic-crypto-mock-logo.svg';
import { fetchDetails } from '../../redux/details/detailsActions';
import { selectCoinDetails, selectDetailsError, selectDetailsIsLoading } from '../../redux/details/detailsSelectors';

const CoinDetails = () => {
  const { id } = useParams();
  const dispatch = useDispatch();
  const coin = useSelector(selectCoinDetails);
  const detailsLoading = useSelector(selectDetailsIsLoading);
  const detailsError = useSelector(selectDetailsError);

  useEffect(() => {
    dispatch(fetchDetails(id));
  }, [id, dispatch]);

  if (detailsLoading) {
    return (
      <>
        <h3>Status</h3>
        <p>Loading...</p>
      </>
    );
  }

  if (detailsError) {
    return (
      <>
        <h4>Error</h4>
        <p>
          <span>{id}</span>
          is not a valid coin.
        </p>
        <p>Please select a Coin from the list.</p>
      </>
    );
  }

  return coin ? (
    <section className="coinDetails">
      <article data-testid={id} className="coin">
        <div className="coinBasics">
          <div className="iconContainer">
            <img data-testid="coinIcon" src={process.env.REACT_APP_API === 'intercept' ? genericCoinIcon : coin.icon} className="coinIcon" alt={coin.name} />
          </div>
          <section className="descriptionSection">
            <h3 data-testid="coinName" className="coinName">{coin.name}</h3>
            <h4 data-testid="coinSymbol" className="sectionCategory symbolCategory">Symbol</h4>
            <p className="sectionData symbolData">{coin.symbol}</p>
          </section>
        </div>
        {/* EOF .coinBasics */}
        <div className="coinDescription">
          <section className="descriptionSection">
            <h3 className="sectionCategory rankCategory">Rank</h3>
            <p data-testid="coinRank" className="sectionData rankData">{coin.rank}</p>
          </section>
          <section className="descriptionSection">
            <h4 className="sectionCategory">Price</h4>
            <p data-testid="coinPrice" className="sectionData">
              {coin.price.toLocaleString(undefined, { maximumFractionDigits: 2 })}
            </p>
            <h4 className="sectionCategory">Volume</h4>
            <p data-testid="coinVolume" className="sectionData">
              {coin.volume.toLocaleString(undefined, { maximumFractionDigits: 2 })}
            </p>
          </section>
          <section className="descriptionSection">
            <h4 className="sectionSubCategory">Price Change</h4>
            <h5 className="sectionTitle">1 hour</h5>
            <p data-testid="coinPrice1h" className="sectionData">
              {coin.priceChange1h.toLocaleString(undefined, { maximumFractionDigits: 2 })}
            </p>
            <h5 className="sectionTitle">1 day</h5>
            <p data-testid="coinPrice1d" className="sectionData">
              {coin.priceChange1d.toLocaleString(undefined, { maximumFractionDigits: 2 })}
            </p>
            <h5 className="sectionTitle">1 week</h5>
            <p data-testid="coinPrice1w" className="sectionData">
              {coin.priceChange1w.toLocaleString(undefined, { maximumFractionDigits: 2 })}
            </p>
          </section>
        </div>
        {/* EOF .coinMore */}
      </article>
    </section>
  ) : (<h3>Select a Coin from the previous list.</h3>);
};

export default CoinDetails;
