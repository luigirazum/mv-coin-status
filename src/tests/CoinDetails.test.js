import { create } from 'react-test-renderer';
import server from './mocks/msw/server';
import renderWithProvider from '../react/utils/renderUtils';
import { mockCoins, mockBitcoin } from './mocks/mockData';
import CoinDetails from '../react/routes/CoinDetails';

const testDetailsBitcoin = {
  details: {
    asset: mockBitcoin,
    status: { type: 'idle' },
    error: null,
  },
};

const testState = {
  coins: {
    assets: [...mockCoins.coins],
    status: { type: 'idle' },
    filter: { active: false, by: '' },
    error: null,
  },
  ...testDetailsBitcoin,
};

const testRouter = {
  routerWrapper: {
    memory: true,
  },
  routes: ['/coins/bitcoin'],
  path: '/coins/:id',
};

let testRender = null;
let container = null;
let component = null;
let getByTestId;

const resetTestEnv = () => {
  container.remove();
  container = null;
  component = null;
};

beforeAll(() => {
  // start API mocking before all tests
  server.listen();
});

beforeEach(() => {
  // setup a DOM element as a render target
  testRender = renderWithProvider(
    <CoinDetails />,
    { preloadedState: { ...testState } },
    { ...testRouter },
  );
  ({
    container, component, getByTestId,
  } = testRender);
  document.body.appendChild(container);
});

afterEach(() => {
  //  cleanup on exiting
  resetTestEnv();
  // reset request handlers at mocked API
  server.resetHandlers();
});

afterAll(() => {
  // closes the mocked API server
  server.close();
});

/** START: CoinDetails component test */
describe('CoinDetails component tests', () => {
  /** START: CoinDetails render at /coins/bitcoin */
  describe('at /coins/bitcoin renders bitcoin currency details', () => {
    test('it\'s id is bitcoin', () => {
      expect(getByTestId('bitcoin')).toBeInTheDocument();
    });
    test('it has a Name', () => {
      expect(getByTestId('coinName')).toBeInTheDocument();
    });
    test('it has a Symbol', () => {
      expect(getByTestId('coinSymbol')).toBeInTheDocument();
    });
    test('it has a Rank', () => {
      expect(getByTestId('coinRank')).toBeInTheDocument();
    });
    test('it has a Price', () => {
      expect(getByTestId('coinPrice')).toBeInTheDocument();
    });
    test('it has a Volume', () => {
      expect(getByTestId('coinVolume')).toBeInTheDocument();
    });
    test('it has a price per hour', () => {
      expect(getByTestId('coinPrice1h')).toBeInTheDocument();
    });
    test('it has a price per day', () => {
      expect(getByTestId('coinPrice1d')).toBeInTheDocument();
    });
    test('it has a price per week', () => {
      expect(getByTestId('coinPrice1w')).toBeInTheDocument();
    });
    test('it has an image icon', () => {
      expect(getByTestId('coinIcon')).toBeInTheDocument();
    });
  });
  /** END: CoinDetails render at /coins/bitcoin */

  /** START: CoinDetails snapshot test */
  describe('CoinDetails component snapshot test', () => {
    test('snapshot renders CoinDetails correctly', () => {
      const tree = create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  /** END: CoinDetails snapshot test */
});
/** END: CoinDetails component test */
