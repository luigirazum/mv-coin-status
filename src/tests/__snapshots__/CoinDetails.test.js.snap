// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinDetails component tests CoinDetails component snapshot test snapshot renders CoinDetails correctly 1`] = `
<section
  className="coinDetails"
>
  <article
    className="coin"
    data-testid="bitcoin"
  >
    <div
      className="coinBasics"
    >
      <div
        className="iconContainer"
      >
        <img
          alt="Bitcoin"
          className="coinIcon"
          data-testid="coinIcon"
          src="generic-crypto-mock-logo.svg"
        />
      </div>
      <section
        className="descriptionSection"
      >
        <h3
          className="coinName"
          data-testid="coinName"
        >
          Bitcoin
        </h3>
        <h4
          className="sectionCategory symbolCategory"
          data-testid="coinSymbol"
        >
          Symbol
        </h4>
        <p
          className="sectionData symbolData"
        >
          BTC
        </p>
      </section>
    </div>
    <div
      className="coinDescription"
    >
      <section
        className="descriptionSection"
      >
        <h3
          className="sectionCategory rankCategory"
        >
          Rank
        </h3>
        <p
          className="sectionData rankData"
          data-testid="coinRank"
        >
          1
        </p>
      </section>
      <section
        className="descriptionSection"
      >
        <h4
          className="sectionCategory"
        >
          Price
        </h4>
        <p
          className="sectionData"
          data-testid="coinPrice"
        >
          27,242.27
        </p>
        <h4
          className="sectionCategory"
        >
          Volume
        </h4>
        <p
          className="sectionData"
          data-testid="coinVolume"
        >
          82,809,392,859.99
        </p>
      </section>
      <section
        className="descriptionSection"
      >
        <h4
          className="sectionSubCategory"
        >
          Price Change
        </h4>
        <h5
          className="sectionTitle"
        >
          1 hour
        </h5>
        <p
          className="sectionData"
          data-testid="coinPrice1h"
        >
          -0.17
        </p>
        <h5
          className="sectionTitle"
        >
          1 day
        </h5>
        <p
          className="sectionData"
          data-testid="coinPrice1d"
        >
          -2.98
        </p>
        <h5
          className="sectionTitle"
        >
          1 week
        </h5>
        <p
          className="sectionData"
          data-testid="coinPrice1w"
        >
          -10.38
        </p>
      </section>
    </div>
  </article>
</section>
`;
